### PR TITLE
log-backup: fix slow flush in GCP

### LINKED ISSUE #13688: pitr backup lag continuously increase in dbaas gcp environment
## Development Task

cc https://github.com/pingcap/tidb/issues/38642

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Close #13688 ; close https://github.com/pingcap/tidb/issues/38642

What's Changed:
This PR used an alternative of the `read_to_end`  to get rid of the performance issue of the origin version of `read_to_end`.

This PR also have enhanced observability for GCS external storage.

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
I have tested read via `read_to_end` and this implementation in a short code snippet, it is about 200x faster (120s vs 600ms) than the origin `read_to_end` when reading a 128MB file.

### Release note 

```release-note
Fixed a bug that may cause high lag of checkpoints in the GCS storage.
```

### CODE REVIEW

**Review by overvenus — APPROVED** [review]

**Comment by overvenus** [comment]
Is it possible to add a benchmark?

**Review by 3pointer — APPROVED** [review]

File: components/cloud/gcp/src/gcs.rs [hunk 1]
```diff
@@ -605,6 +630,83 @@ mod tests {
...
+    #[bench]
+    fn bench_read_to_end(b: &mut test::Bencher) {
+        b.iter(|| {
+            let mut v = [0; BENCH_READ_SIZE];
+
+            let mut r = ThrottleRead::new(Cursor::new(&mut v));
+            let mut dst = Vec::with_capacity(BENCH_READ_SIZE);
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .build()
+                .unwrap();
```
Comment (overvenus) [inline components/cloud/gcp/src/gcs.rs:686]: They are noises for a benchmark of `read_to_end`, they should be moved out the `b.iter()`.

```rust
let mut dst = Vec::with_capacity(BENCH_READ_SIZE);
let rt = tokio::runtime::Builder::new_current_thread()
    .build()
    .unwrap();
let mut v = [0; BENCH_READ_SIZE];
b.iter(|| {
    let mut r = ThrottleRead::new(Cursor::new(&mut v));
    dst.clear();
    rt.block_on(r.read_to_end(&mut dst)).unwrap();
    assert_eq!(dst.len(), BENCH_READ_SIZE)
})
```

File: components/cloud/gcp/src/gcs.rs [hunk 2]
```diff
@@ -605,6 +630,83 @@ mod tests {
...
+    #[bench]
+    fn bench_manual_read_to_end(b: &mut test::Bencher) {
+        b.iter(|| {
+            let mut v = [0; BENCH_READ_SIZE];
+
+            let r = ThrottleRead::new(Cursor::new(&mut v));
+            let mut dst = Vec::with_capacity(1024);
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .build()
+                .unwrap();
```
Comment (overvenus) [inline components/cloud/gcp/src/gcs.rs:703]: See the above comment.

**Comment by YuJuncen** [comment]
(Manually removed [do-not-merge/needs-linked-issue](https://github.com/tikv/tikv/labels/do-not-merge%2Fneeds-linked-issue) because it has already linked to a issue in the pingcap/tidb repo)

**Pull request merged**